### PR TITLE
[Bugfix][Kernel] GDN readout: honor fp32 SSM state precision to avoid fp16 overflow → NaN

### LINKED ISSUE #54308: GDN / Qwen3-Next decode degenerates to a single repeated token on non-Blackwell GPUs with an fp32 SSM cache
Running Qwen3-Next (GDN) on 4x MI50 (gfx906, ROCm) with `--mamba-ssm-cache-dtype float32`, decode eventually collapses to a single token repeated forever (`!!!!!!...`) once a conversation gets long enough. Greedy or low temp, doesn't matter — short prompts are fine, it only shows up after the recurrent state has grown over a lot of context.

Took a while to track down. It's in `fused_sigmoid_gating_delta_rule_update`: the readout `o = state @ q` is accumulated in fp32 (`b_o` is `tl.float32`), but the output tensor is allocated in the activation dtype —

```python
o = q.new_empty(NK, *v.shape)   # inherits q's dtype, i.e. fp16
```

The GDN state is an unnormalized accumulator (only normalized at the readout, by `RMSNormGated`), so with an fp32 state cache the readout runs past fp16's 65504, overflows to `inf` at the store, `RMSNormGated` turns `inf/inf` into `NaN`, and decode repeats one token from there on. Blackwell doesn't hit it because the `fi`/`cutedsl` backends do the readout in fp32 — it's specific to the Triton path everything else uses.

Minimal kernel repro (fails on main, passes once the output honors the state dtype):

```python
import torch
from vllm.platforms import current_platform
from vllm.utils.torch_utils import set_random_seed
from vllm.model_executor.layers.fla.ops import fused_sigmoid_gating_delta_rule_update

DEVICE = current_platform.device_type
set_random_seed(0)
B, T, H, HV, K, V = 1, 1, 4, 4, 64, 128
dt = torch.float16
q = torch.randn(B, T, H, K, device=DEVICE, dtype=dt)
k = torch.randn(B, T, H, K, device=DEVICE, dtype=dt)
v = torch.randn(B, T, HV, V, device=DEVICE, dtype=dt)
a = torch.randn(B, T, HV, device=DEVICE, dtype=dt)
b = torch.randn(B, T, HV, device=DEVICE, dtype=dt)
A_log = torch.randn(HV, device=DEVICE, dtype=torch.float32)
dt_bias = torch.randn(HV, device=DEVICE, dtype=torch.float32)
state = torch.randn(B, HV, V, K, device=DEVICE, dtype=torch.float32) * 1e6  # fp32, > fp16 max

o, _ = fused_sigmoid_gating_delta_rule_update(
    A_log=A_log, a=a, b=b, dt_bias=dt_bias, q=q, k=k, v=v,
    initial_state=state.clone(), inplace_final_state=False,
    use_qk_l2norm_in_kernel=True,
)
print("non-finite:", int((~torch.isfinite(o)).sum()), "max|o|:", o.float().abs().max().item())
# main: non-finite: 107  max|o|: inf
```

Same materialize-in-activation-dtype pattern is also in the chunk/prefill kernels (`chunk_o`, `chunk_delta_h`, `wy_fast`) and the `core_attn_out` buffer, but the decode-update kernel above is the one that produces the actual `!!!!` collapse.

Fix is a one-liner — allocate the output in the state's precision, conditional on `initial_state.dtype` so the fp16 path is untouched. Opened #54146 with a regression test.

### Environment
- 4x AMD MI50 (gfx906), ROCm 6.3.3
- vLLM 0.27.2rc1
- Qwen3-Next / Qwen3.5 hybrid (GDN), AWQ, TP=4
- `--dtype float16 --mamba-cache-dtype float32 --mamba-ssm-cache-dtype float32`

### PR DESCRIPTION
**Pull request by JackDanger**
## Purpose

Closes #54308.

On non-Blackwell GPUs, gated-delta-net (Qwen3-Next / GDN) decode degenerates to a single infinitely-repeated token when the SSM state cache is fp32 (`--mamba-ssm-cache-dtype float32`). Once the recurrent state grows large over long context, greedy/low-temperature decoding emits one token (e.g. `!!!!…`) forever.

## Root cause

`fused_sigmoid_gating_delta_rule_update` accumulates the readout `o = state @ q` in fp32 (`b_o` is `tl.float32`) but materialized the output tensor in the input/activation dtype:

```python
o = q.new_empty(NK, *v.shape)  # inherits q's dtype, e.g. fp16
```

The GDN recurrent state is an *unnormalized* accumulator: it grows over long, low-decay context and is normalized only at the readout by `RMSNormGated`. With an fp32 state cache the readout exceeds fp16's 65504 range and overflows to `±inf` at the store; `RMSNormGated` then produces `inf / inf = NaN`, and decode collapses to a single repeated token. Blackwell does not hit this because its `fi` / `cutedsl` GDN backends run the readout in fp32; the Triton path here is the one every non-Blackwell GPU takes.

## Fix

Materialize the output in the state's precision — fp32 when `initial_state` is fp32, unchanged otherwise. The fp16 path is byte-for-byte identical, so tensor-core / NVIDIA behavior is unaffected by construction.

The same materialize-in-activation-dtype pattern exists in the chunk/prefill kernels (`chunk_o`, `chunk_delta_h`, `wy_fast`) and the `core_attn_out` buffer. This PR fixes the decode-update kernel, which is the one that produces the observed decode degeneration; the chunk-path kernels are a straightforward follow-up.

## Test Plan

Adds `test_fused_sigmoid_gating_readout_finite_under_large_fp32_state` to `tests/kernels/test_fused_sigmoid_gating_delta_rule.py`: drives the decode kernel with fp16 activations and a large fp32 `initial_state` (`1e6`, past fp16 max) and asserts the readout is finite and preserves the true magnitude (not saturated to fp16 max).

## Test Result

- Before: 107 non-finite entries, `max|o| = inf`.
- After: finite, `max|o| > 1e4`.

Validated on AMD MI50 (gfx906). The change is gated on `initial_state.dtype`, so the fp16 path — and thus NVIDIA behavior — is unchanged.

### CODE REVIEW

**Comment by JartX** [comment]
Hi @JackDanger this PR has been helpful, but its incomplete for me, can you check my commit and take my second patch on your pr please?  think that  both should be merged, on this PR, to estabilice the Qwen3.5 arch models many thanks
